### PR TITLE
TOOLS-2357 add 'cargo clippy' to eng.git 'make check-rust'

### PR DESCRIPTION
**Pull request by trentm**

### CODE REVIEW

**Comment by trentm** [comment]
@kellymclaughlin @rjloura @rhb2 ^^ thoughts on a way (or make it the default?) to have `cargo clippy` be invoked with args to make its warnings be *failures*?  I suspect if just warnings, they will largely be ignored in automated builds (the build output log has enough noise that finding the warnings isn't so easy).

**Comment by papertigers** [comment]
I kind of like the default behavior being to run clippy since it's suggestions are usually super helpful.  Since most clippy warnings can be disabled in the code if one feels strongly about something I don't see it as a big deal.

Maybe we should make the default to run clippy and instead go with something like `RUST_CLIPPY_CLEAN=false` to disable it for a repo.  Thoughts @kellymclaughlin ?

**Comment by kellymclaughlin** [comment]
@trentm So my opinion on this is we lean toward being more stringent by default and treat warnings as errors. I think this will encourage people to examine the warnings and either make the suggested changes or explicitly decide that the lint is not suitable and add appropriate annotations to indicate the decision. I think doing this will increase the uniformity of our code and hopefully help guide us toward good patterns. I have only had a very few cases where I did not want to adopt the suggestion from clippy and in those cases it wasn't difficult to add the needed annotations to ignore those.

**Comment by kellymclaughlin** [comment]
Sorry, I totally misunderstood what was being asked in my last response. 

I think we can invoke clippy and set the lint groups to `deny` to get the warnings as errors behavior. Something like:

```
cargo clippy -D clippy::all -D clippy::pedantic ...
```

**Comment by kellymclaughlin** [comment]
Maybe just the `clippy::all` group is what we should use. This represents all the lints that are enabled by default. We can leave it up to each project if they want to run the `pedantic` checks themselves.

**Comment by trentm** [comment]
https://github.com/rust-lang/rust-clippy#usage has these examples:

```
  # if you want the build job to fail when encountering warnings, use
  - cargo clippy -- -D warnings
  # in order to also check tests and non-default crate features, use
  - cargo clippy --all-targets --all-features -- -D warnings
```

I am personally confused by clippy's "options" vs `<opts>`:

```
$ cargo clippy --help
Checks a package to catch common mistakes and improve your Rust code.

Usage:
    cargo clippy [options] [--] [<opts>...]

Common options:
    -h, --help               Print this message
    -V, --version            Print version info and exit

Other options are the same as `cargo check`.

To allow or deny a lint from the command line you can use `cargo clippy --`
with:

    -W --warn OPT       Set lint warnings
    -A --allow OPT      Set lint allowed
    -D --deny OPT       Set lint denied
    -F --forbid OPT     Set lint forbidden

You can use tool lints to allow or deny lints from your code, eg.:

    #[allow(clippy::needless_lifetimes)]
```

**Comment by trentm** [comment]
> We can leave it up to each project if they want to run the `pedantic` checks themselves.

That would suggest that a `RUST_CLIPPY_ARGS` (or similar) var in the eng.git Makefiles would be useful. I'll add something like that.

That would also give an "out" if our adding this force an existing rust project to have to cope with clippy *right now*.

**Comment by rhb2** [comment]
> Maybe just the `clippy::all` group is what we should use. This represents all the lints that are enabled by default. We can leave it up to each project if they want to run the `pedantic` checks themselves.

Personally, I don't mind if clippy plays hard ball during build time (i.e. pedantic).  Most of the things that I've observed it warning about are usually for a good reason even if not immediately dangerous for the application.  It would also prevent situations where we binge clippy clean which does occasionally happen in the rebalancer, usually accompanied by some unrelated change which makes the review process that much more complicated.  As a matter of fact, the other thing that I'd be in favor of adding a call to "rustfmt --check" which would return 1 if Rust would make any formatting changes, otherwise 0.  I'd be in favor of adding it for the same reasons as the aggressive clippy -- when someone makes a change and decides that today is the day that they'd like to run cargo fmt, the result is a cr (or I guess pr now) where it's difficult to know what's part of the real change and what isn't.

**Comment by rjloura** [comment]
There's been one or two times where I've disagreed with clippy.  It would be good if we could have clippy ignore a line in source in much the same way as we can use comments for lint.

**Comment by trentm** [comment]
All, the latest commit does:
- plays "hardball" by defaulting to clippy::all and pedantic
- adds a `RUST_CLIPPY_ARGS` var that allows overriding the args in a repo's Makefile

Testing notes at: https://jira.joyent.us/browse/TOOLS-2357?focusedCommentId=275220&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-275220

File: tools/mk/Makefile.targ [hunk 1]
```diff
@@ -284,10 +298,13 @@ JSSTYLE_TARGET = $(if $(JSSTYLE_FILES), check-jsstyle)
 check-jsstyle:  $(JSSTYLE_EXEC)
 	$(JSSTYLE) $(JSSTYLE_FLAGS) $(JSSTYLE_FILES)
 
+# We run `cargo clean` before `cargo clippy` because, at least in Rust v1.35
+# clippy with not check already compiled files.
```
Comment (timfoster) [inline tools/mk/Makefile.targ]: Are we missing words in this sentence somewhere?

**Pull request merged**